Don't panic if message.Data can't be converted to json

tinygrasshopper opened the pull request:
Scenario:
When LogFormat.Data is assigned an object that cannot be converted to a json, like channel, complex, and function values.

Current behaviour:
It panics 

Expected behaviour:
Should not panic 

What should be logged if the json serialization fails is a bit of a puzzle. If the json serialization fails, it can be only because of the Data attribute as all of the other fields on LogFormat can always be encoded into json. I think the options are:
1. Don't log anything
2. Log the message without the `Data` attribute.
3. Log the message by replacing data with the error that occurred while marshalling data.

Option 3 is chosen as its the most verbose, but I think it might be confusing for the reader, as the message can be mistaken for an error being logged, rather than a error while logging.

Thoughts?

ematpl commented:
Thanks, @tinygrasshopper, I also think it would be beneficial for the lager library to be more resilient to these kinds of marshalling errors. Given that the only possible serialization errors for the `LogFormat` struct would come from one of the values in the `Data` map, I think it would be even more appropriate to replace only the value of those data entries with the error that comes from failing to serialize that value. Then log consumers would see as much as possible about their log-message annotations.

I think it would also be helpful to wrap that error message in a message from the lager library that clearly indicates that that replacement has taken place, so that component operators can easily scan their logs for those errors and fix them in the upstream codebase.

Would you be willing to make those improvements to the PR?

Thanks again,
Eric, CF Runtime Diego PM

tinygrasshopper commented:
@ematpl, yes I am open to changing the PR. I agree that we should make the error more explicit so that it looks like a lager error rather than an application error. I have changed the message to reflect that.

The JSON message written when there is a serialisation error is:
`{"data":{"lager serialisation error": "json: unsupported type: func()"}}`

Do you think this is enough to disambiguate this error from any application error that might be logged?

tinygrasshopper commented:
hey @ematpl any updates on this?

ematpl commented:
Hey, @tinygrasshopper, sorry for the delay. That's probably fine as a way for lager to degrade gracefully in the face of a serialization error. Given that the only way a serialization error can occur is from one of the values in the lager.Data map, though, I was thinking that even better behavior would be to scope the error to that key-value pair in the output. For example,

```
logger.Info("important-event", lager.Data{"status": "hunky-dory", "metadata": func() string { return "nah, dawg" }})
```

would result in a log line like

```
{"message":"important-event","log_level":1,"timestamp":"","source":"","data":{"status":"hunky-dory","metadata":"lager serialization error: json: unsupported type: func() string","session":"12.98.1"}}
```

This would preserve as much of the log-line metadata as possible (including the session sequence) while indicating that the serialization error arose from one specific key in the attached Data. What do you think?

Thanks,
Eric

andrew-edgar commented:
Hello @tinygrasshopper ,

What do you think of the last comments from @ematpl?  We would like to get your feedback and would like to know if you are planning to update the PR to address the conversation.

Thanks,

Andrew Edgar & John Shahid

jvshahid commented:
bump

tinygrasshopper commented:
Sorry @andrew-edgar @jvshahid this one fell off my radar.

I agree with @ematpl solution in theory, but can't think of a clean way of implementing it. 
I think essentiality we would need a replacement for `json.Marshal` which rather than returning an error, encodes the error to the json document with a prefix, which would be quite an undertaking.

@ematpl are you ok with accepting this PR as a way to stop the panics, and improve the logging output later?

ematpl commented:
Hmm, I think a custom `MarshalJSON` method on `lager.Data` would work. Here's an example (also at https://play.golang.org/p/8Z6qM_LKQ):

```
package main

import (
	"encoding/json"
	"fmt"
)

type Data map[string]interface{}

func (d Data) MarshalJSON() ([]byte, error) {
	im := make(map[string]*json.RawMessage)

	for k, v := range d {
		p, err := json.Marshal(v)
		if err != nil {
			p, err = json.Marshal("lager serialization error: " + err.Error())
			if err != nil {
				// wat
				panic(err)
			}
		}
		j := json.RawMessage(p)
		im[k] = &j
	}

	return json.Marshal(im)
}

func main() {
	data := Data{"foo": "bar", "baz": 1, "no": func() string { return "dawg" }, "nil": nil}
	dataPayload, err := json.Marshal(data)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(dataPayload))
}
```

I haven't checked how much less efficient than the standard library it is, though.

tinygrasshopper commented:
@ematpl Good one, I was thinking you where implying a custom recursive decent serialiser which will not serialize the erring attribute, restricting it to one level makes much more sense. 

Have updated the code, with your snippet.

ematpl commented:
Thanks for the update, @tinygrasshopper. @bdshroyer did some benchmarking of the more granular error-handling on the [wip-lager-benchmark branch](https://github.com/cloudfoundry/lager/tree/wip-lager-benchmark), and the [results here](https://docs.google.com/spreadsheets/d/1kNg4Xobg3TE1DpJ9TsDOeriqN506sRX7QSrkriyCXXg/edit) show a 40% to 75% performance hit. Given that we expect never to encounter this case in production, that's a pretty high cost to pay. Unless the performance for the current approach can be improved so that it's only a negligible increase in time/CPU/memory, we should go back to the coarser approach (if anything).

Thanks again,
Eric

tinygrasshopper commented:
Hi Eric,

I can see how the custom JSON formatter approach will incur a hit on the performance. 

Can we revert back to the simpler approach of handling JSON errors, and print a lager error message. See https://github.com/cloudfoundry/lager/pull/20/commits/526cc916a06edc0b21003284f76ff45120c160b9

Leaving it in the state that it is in, where it panics if it can't format the message is not great.

Thoughts?

ematpl commented:
@tinygrasshopper, sure, but let's go with the more specific `lager_serialization_error` key you introduced in https://github.com/cloudfoundry/lager/pull/20/commits/23454dcd4fc21e7f5fa5379f38d5253df5ff29f0, instead of the generic `error` key that could be conflated with an error-level log message.

I'm still concerned that in the error case the log wouldn't show enough information about the data to determine why the serialization error occurred. One thing you could also add in would be to dump the data safely via `fmt.Sprintf("%#v", data)`. Here's an analogous snippet for that:

```
package main

import (
	"encoding/json"
	"fmt"
)

type Data map[string]interface{}

func (d Data) MarshalJSON() ([]byte, error) {
	payload, err := json.Marshal(map[string]interface{}(d))
	if _, ok := err.(*json.UnsupportedTypeError); ok {
		return json.Marshal(map[string]string{
			"lager_serialization_error": err.Error(),
			"data_dump":                 fmt.Sprintf("%#v", d),
		})
	}

	return payload, err
}

func main() {
	data := Data{"foo": "bar", "baz": 1, "no": func() string { return "dawg" }, "nil": nil}
	dataPayload, err := json.Marshal(data)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(dataPayload))
}
```

I don't expect that would have a substantial performance hit on the non-error case because it would incur only an extra type assertion, but if you try that approach out would you mind running those comparative benchmarks for it (or even putting together a `Measure` ginkgo test for it)?

Thanks,
Eric

bdshroyer commented:
Hi @tinygrasshopper,

Any updates on this?


Thanks,

Brandon

tinygrasshopper commented:
Reverted to the previous implementation, added the data_dump field. 

Without `data_dump` for writing 5000 lines each run. 
```
    Ran 10 samples:
    runtime:
      Fastest Time: 0.335s
      Slowest Time: 0.394s
      Average Time: 0.350s ± 0.017s
```


With `data_dump` for writing 5000 lines each run. 
```
    Ran 10 samples:
    runtime:
      Fastest Time: 0.477s
      Slowest Time: 0.524s
      Average Time: 0.496s ± 0.014s
```

So adding the `data_dump` would make it ((0.496-0.350)/0.350)*100 = 41.7 % Slower.
Thoughts @ematpl?

ematpl commented:
Hey, @tinygrasshopper, that slowdown with data_dump is incurred only when there's a serialization error, right? If so, that seems fine to me. Would you mind squashing those intermediate commits down to a single one with the final set of changes?

Thanks,
Eric

tinygrasshopper commented:
@ematpl that is right. Changes have been squashed

ematpl commented:
Looks good, @tinygrasshopper! Thanks!
